upt_macports/upt_macports.py

```diff
@@ -176,6 +180,44 @@ def create_package(self, upt_pkg, output=None):
             pkg_cls = pkg_classes[upt_pkg.frontend]
         except KeyError:
             raise upt.UnhandledFrontendError(self.name, upt_pkg.frontend)
-
         packager = pkg_cls()
         packager.create_package(upt_pkg, output)
+
+    def pkg_exist(self, pkg_name, frontend):
+        prefix = {
+            'pypi': 'py-',
+            'rubygems': 'rb-',
+            'cpan': 'p5-',
+        }
+        port_name = prefix.get(self.upt_pkg.frontend, '') + self.upt_pkg.name
+        self.logger.info(f'Checking MaPorts tree for port {port_name}')
+        cmd = f'port info {port_name}'
+        port = os.popen(cmd).read()
+        self.logger.info(f'Current Upstream version is {self.upt_pkg.version}')
+        if port == '':
+            self.logger.info(f'{port_name} not found in MacPorts tree')
+            self.curr_ver = ''
+            return False
+        else:
+            self.curr_ver = port.split('@')[1].split()[0].split('_')[0]
+            self.logger.info(
+                f'Current MacPorts Version for {port_name} is {self.curr_ver}')
+            return True
+
+    def check_needed(self, upt_pkg, constraint):
+        self.upt_pkg = upt_pkg
+        if self.pkg_exist(upt_pkg.name, upt_pkg.frontend):
+            curr_ver = self.curr_ver
+            upstream_ver = self.upt_pkg.version
+            # TODO: Add Decision Logic
+            self.logger.info(f'Constraint is {constraint}')
+            if curr_ver < upstream_ver:
+                self.logger.info(
+                    'Creating Portfile as new version is available')
+                return True
```

Comment:
You are right that this is the scenario where we would update the Portfile but that is only if the `--update` is passed and as of now even if we don't have update working, it's like if the upstream has updated and the online MacPorts tree has older version, then we generate a new Portfile but if theres a local Portfile then it will anyways throw an error while writing.

So I think we should keep it as `True` as even after update is ready but for some reason user wants just the "pure" upt output then this will be operational.

Reply:
I don't think I agree with that ;) At least not if we don't add additional checks with respect to the constraints. If a package ```py-foo``` is happy with whatever version of ```py-bar```, then even if there is a newer version of ```py-bar``` available upstream we should not generate a new Portfile (or probably even update it unconditionally once that feature is available). There might be valid reason for keeping ```py-bar``` at a lower version because of requirements of other packages for example.

So I think we should check whether the available package would fulfill the ```constraints``` and if that's the case we just log "FYI: there is a newer version available upstream" but don't do anything else. If the version is too low then we should certainly log that and only in that case we could consider to return ```True```. But even there I would be in favor of not doing that right now as this would only work properly anyway once the "update" feature is in place. 

@Steap - opinions / insights?

Reply:
I think you want to keep commits as small as possible, so you should not update without a valid reason. Think of a Python package that uses pytest, for instance. If you upgrade pytest and/or its plugins, you might break unrelated pieces of software. 

Therefore I agree with @reneeotten: only upgrade when required.